Add release notes for screen-space reflections (#1442)
bevyengine/bevy-website: release-content/0.14/release-notes/13418_Implement_optin_sharp_screenspace_reflections_for_the_defe.md

Our initial implementation is relatively minimal, to provide a flexible base to build on, but is based on the production-quality [raymarching code by Tomasz Stachowiak](https://gist.github.com/h3r2tic/9c8356bdaefbe80b1a22ae0aaee192db), one of the creators of the indie darling Bevy game [Tiny Glade](https://store.steampowered.com/app/2198150/Tiny_Glade/).
As a result, there are a few caveats to bear in mind:

1. Currently, this feature is built on top of the deferred renderer and is currently only supported in that mode. Forward screen-space reflections are possible albeit uncommon (though e.g. Doom Eternal uses them); however, they require tracing from the previous frame, which would add complexity. This patch leaves the door open to implementing SSR in the forward rendering path but doesn't itself have such an implementation.
2. Screen-space reflections aren't supported in WebGL 2, because they require sampling from the depth buffer, which `naga` can't do because of a bug (sampler2DShadow is incorrectly generated instead of sampler2D; this is the same reason why depth of field is disabled on that platform).

> 3. No temporal filtering or roughness blurring is performed at all. For this reason, SSR currently only operates on very shiny surfaces.

Reviewer: ```suggestion
3. No temporal filtering or roughness blurring is performed at all. For this reason, SSR currently only operates on very low-roughness / smooth surfaces.
```

<!-- Implement opt-in sharp screen-space reflections for the deferred renderer, with improved raymarching code. -->
<!-- https://github.com/bevyengine/bevy/pull/13418 -->

{{ compare_slider(
    path="POST_PATH",
    left_title="No SSR",
    left_image="no_ssr.png",
    right_title="SSR",
    right_image="ssr.png"
) }}

> [Screen-space reflections](https://lettier.github.io/3d-game-shaders-for-beginners/screen-space-reflection.html) approximate real-time reflections by raymarching through the depth buffer and copying samples from the final rendered frame.

Reviewer 2: Nit: We could introduce the `SSR` acronym here for those unfamiliar (myself lol)
```suggestion
[Screen-space reflections](https://lettier.github.io/3d-game-shaders-for-beginners/screen-space-reflection.html) (SSR) approximate real-time reflections by raymarching through the depth buffer and copying samples from the final rendered frame.
```

Our initial implementation is relatively minimal, to provide a flexible base to build on, but is based on the production-quality [raymarching code by Tomasz Stachowiak](https://gist.github.com/h3r2tic/9c8356bdaefbe80b1a22ae0aaee192db), one of the creators of the indie darling Bevy game [Tiny Glade](https://store.steampowered.com/app/2198150/Tiny_Glade/).
As a result, there are a few caveats to bear in mind:

1. Currently, this feature is built on top of the deferred renderer and is currently only supported in that mode. Forward screen-space reflections are possible albeit uncommon (though e.g. Doom Eternal uses them); however, they require tracing from the previous frame, which would add complexity. This patch leaves the door open to implementing SSR in the forward rendering path but doesn't itself have such an implementation.

> 2. Screen-space reflections aren't supported in WebGL 2, because they require sampling from the depth buffer, which `naga` can't do because of a bug (sampler2DShadow is incorrectly generated instead of sampler2D; this is the same reason why depth of field is disabled on that platform).

Reviewer 2: Nit: Is there a tracking issue for this bug? Could be worth linking to it for reference (and in the off chance someone sees this and decides to tackle it themselves)?

Author: I can't find this after quiet a while looking. I'll add a link if it turns up.

> 2. Screen-space reflections aren't supported in WebGL 2, because they require sampling from the depth buffer, which `naga` can't do because of a bug (sampler2DShadow is incorrectly generated instead of sampler2D; this is the same reason why depth of field is disabled on that platform).

Reviewer 2: Nit: Should these be rendered as code?
```suggestion
2. Screen-space reflections aren't supported in WebGL 2, because they require sampling from the depth buffer, which `naga` can't do because of a bug (`sampler2DShadow` is incorrectly generated instead of `sampler2D`; this is the same reason why depth of field is disabled on that platform).
```

> 3. No temporal filtering or blurring is performed at all. For this reason, SSR currently only operates on very low-roughness / smooth surfaces.

Reviewer 2: Question for myself: is spaces before and after a `/` part of our "style guide"? Should I be preferring this style when writing my release notes as well?

Author: Man I have *no* idea  / feelings about this question.

4. We don't perform acceleration via the hierarchical Z-buffer and reflections are traced at full resolution. As a result, you may notice performance issues depending on your scene and hardware.

> ![A cube with the Bevy logo on all 6 sides sits on top of rippling water. Sharp reflections are visible, distorted by the uneven surface of the water.](screen_space_reflections.png)

Reviewer 2: Now _that's_ an alt text!

Author: It's important to convey the full emotion and meaning of the image to blind users!

To add screen-space reflections to a camera, insert the [`ScreenSpaceReflectionsSettings`] component.

> In addition to [`ScreenSpaceReflectionsSettings`], [`DepthPrepass`] and [`DeferredPrepass`] must also be present for the reflections to show up.

Reviewer 2: Nit: I feel like I ask this all the time, but do we prefer Oxford comma or not? I'll try to remember this time lol
```suggestion
In addition to [`ScreenSpaceReflectionsSettings`], [`DepthPrepass`], and [`DeferredPrepass`] must also be present for the reflections to show up.
```

Author: We prefer it.

Conveniently, the [`ScreenSpaceReflectionsBundle`] bundles these all up for you!
While the [`ScreenSpaceReflectionsSettings`] comes with sensible defaults, it also contains several settings that artists can tweak.

[`ScreenSpaceReflectionsBundle`]: https://dev-docs.bevyengine.org/bevy/pbr/struct.ScreenSpaceReflectionsBundle.html
[`ScreenSpaceReflectionsSettings`]:https://dev-docs.bevyengine.org/bevy/pbr/struct.ScreenSpaceReflectionsSettings.html
[`DepthPrepass`]: https://dev-docs.bevyengine.org/bevy/core_pipeline/prepass/struct.DepthPrepass.html

> [`DeferredPrepass`]: https://dev-docs.bevyengine.org/bevy/core_pipeline/prepass/struct.DeferredPrepass.html

Reviewer 2: Question: do these links get updated once the release is out?

Author: Yep, that's going to be a global find-and-replace during the editing pass.
